### PR TITLE
Show stack traces in CoqIDE error dialog when relevant

### PR DESCRIPTION
**Pull request by jfehrle**
Make the former "coqidetop died" dialog useful for reporting and debugging issues:
- always record stack trace in CoqIDE
- display the stack trace for exceptions that previously caused "coqidetop died" failures
- correctly distinguish death of the coqidetop process (labelled "coqidetop died") from exceptions within CoqIDE

Exceptions generated in GTK callbacks don't by default give a useful stack trace or show a message other than on the console.  However, adding `exc @@ fun () ->` at the beginning of any callback will enable showing a useful stack trace in a dialog.  I chose not to instrument every callback right now since that's ~200 places to modify.

The second commit shows examples that force exceptions for CTRL-End, CTRL-Up and fetching the goal info after processing a sentence.  _**The second commit should not be merged. I'll remove it after you've examined it.**_

The question dialog doesn't support copying the text of the exception, so users will need to submit screen shots.  Hopefully it's very rare to get such errors.

image

### CODE REVIEW

**Comment by Zimmi48** [comment]
> The second commit shows examples that force exceptions for CTRL-End, CTRL-Up and fetching the goal info after processing a sentence. _**The second commit should not be merged. I'll remove it after you've examined it.**_

I've added a "needs: fixing" label so that this is not forgotten (we could imagine using a dedicated "needs: " label, but currently this was the closest thing that was available).

**Comment by Alizter** [comment]
Can you put the backtrace in a scrollable textbox? An added "copy" button would also be welcome.

**Comment by jfehrle** [comment]
While those would be nice touches, I think they're not essential.  I've found that making some of these apparently trivial interface changes end up taking hours to get right.  Partly just the nature of GUI development, but GTK bugs and limited documentation (i.e., GTK is only documented for the C interface) don't help. Nor am I making similar changes often enough to become really efficient at making them.  There are many other things to fix/improve in CoqIDE and Coq that are more important.  Developer time is always limited.

**Comment by SkySkimmer** [comment]
>I've added a "needs: fixing" label so that this is not forgotten (we could imagine using a dedicated "needs: " label, but currently this was the closest thing that was available).

This fits more into needs:progress imo

**Comment by jfehrle** [comment]
More like needs:review.

**Comment by ppedrot** [comment]
This is not a bugfix so it doesn't qualify for a 8.15 backport. The 8.16 branch is around the corner, and this is neither critical nor has not been reviewed yet, so I'm postponing it to 8.17.

**Comment by SkySkimmer** [comment]
ping @coq/coqide-maintainers

**Comment by ppedrot** [comment]
This PR doesn't seem to be ready for review, there are uses of `List.hd []` in the code to trigger backtraces for debug purposes.

**Comment by SkySkimmer** [comment]
That's the example commit not for debugging

**Comment by jfehrle** [comment]
Note that the stack traces will only cover the OCaml routines called since the last callback from GTK.  So to get good coverage, that means adding `exc @@ fun () ->` boiler plate for every menu action routine, etc.  IIRC that's more than 100 places.  I didn't add all those yet (only a couple in the 2nd commit as examples) which I perhaps should.

The "Please report at ..." may give us quite a few bug reports that may not have enough information to resolve.  If that's the case, we can adjust the message in the next release.

Off topic: Currently you can only ignore the error or exit CoqIDE.  It would be preferable to have an option to just drop a specific buffer if we can tell which one caused the problem.

File: ide/coqide/coqOps.ml [hunk 1]
```diff
@@ -424,6 +424,7 @@ object(self)
     Coq.bind call begin function
     | Fail x -> self#handle_failure_aux ~move_insert x
     | Good goals ->
+      let _ = List.hd [] in
```
Comment (SkySkimmer) [inline ide/coqide/coqOps.ml:427]: wtf

**Comment by SkySkimmer** [comment]
With no progress for years let's close

**Pull request closed without merging**